src/input/type_record.cc

```diff
@@ -184,25 +169,25 @@ void Record::make_copy_keys(Record &origin) {
 }
 
 
-void Record::make_copy_keys_all() {
-	for(auto &ptr : data_->copy_from_ptr) {
-		ASSERT( ptr && TypeBase::was_constructed( ptr ), "Invalid pointer to source record for copy keys operation.\n");
-		Record tmp(*ptr);
-		make_copy_keys(tmp);
-		ptr = NULL;
-	}
-	data_->copy_from_ptr.clear();
-}
-
-
 
 Record &Record::derive_from(AbstractRecord &parent) {
-	ASSERT( ! data_->p_parent_ && ! data_->parent_ptr_ , "Record has been already derived.\n");
-    if (TypeBase::was_constructed(&parent)) {
-        data_->parent_ptr_=boost::make_shared<AbstractRecord>(parent);
-        data_->p_parent_ = NULL;
-    } else { //postponed
-        data_->p_parent_ = &parent;
+	//ASSERT( ! data_->parent_ptr_.size() , "Record has been already derived.\n");
+	ASSERT( parent.is_closed(), "Parent AbstractRecord '%s' must be closed!\n", parent.type_name().c_str());
+	ASSERT( data_->keys.size() == 0 || (data_->keys.size() == 1 && data_->keys[0].key_ == "TYPE"),
+			"Derived record '%s' can have defined only TYPE key!\n", this->type_name().c_str() );
+
+	bool contains = false;
+	for (auto it = data_->parent_ptr_.begin(); it != data_->parent_ptr_.end(); ++it) {
+	    if ((*it)->type_name() == parent.type_name()) {
```

Comment:
I suggest to rather keep all pointers and make this list unique during finish. 
Then during finish we have to use hash of the parent abstract, otherwise we are not able to distinguish abstracts comming from common generic type.